cmd/lotus-shed/fip-0036.go

```diff
@@ -0,0 +1,565 @@
+		// a map of accounts to every msig that they are a signer of
+		accountsToMultisigs := make(map[address.Address][]address.Address)
+		// a map of multisigs to some info about them for quick lookup
+		msigActorsInfo := make(map[address.Address]msigBriefInfo)
+		// a map of accounts to every miner that they are an owner of
+		ownerMap := make(map[address.Address][]address.Address)
+		// a map of accounts to every miner that they are a worker of
+		workerMap := make(map[address.Address][]address.Address)
+		// a map of miners to some info aboout them for quick lookup
+		minerActorsInfo := make(map[address.Address]minerBriefInfo)
+		// a map of client addresses to deal data stored in proposals
+		clientToDealStorage := make(map[address.Address]abi.StoragePower)
+
+		fmt.Println("iterating over all actors")
+		count := 0
+		err = st.ForEach(func(addr address.Address, act *types.Actor) error {
+			if count%200000 == 0 {
+				fmt.Println("processed ", count, " actors building maps")
+			}
+			count++
+			if builtin.IsMultisigActor(act.Code) {
+				ms, err := multisig.Load(store, act)
+				if err != nil {
+					return fmt.Errorf("load msig failed %v", err)
+
+				}
+
+				// TODO: Confirm that these are always ID addresses
+				signers, err := ms.Signers()
+				if err != nil {
+					return xerrors.Errorf("fail to get msig signers", err)
+				}
+				for _, s := range signers {
+					signerId := lookupId(s)
+					if m, found := accountsToMultisigs[signerId]; found { //add msig id to signer's collection
+						m = append(m, addr)
+						accountsToMultisigs[signerId] = m
+					} else {
+						n := []address.Address{addr}
+						accountsToMultisigs[signerId] = n
+					}
+				}
+
+				locked, err := ms.LockedBalance(abi.ChainEpoch(height))
+				if err != nil {
+					return xerrors.Errorf("failed to compute locked multisig balance: %w", err)
+				}
+
+				threshold, _ := ms.Threshold()
+				info := msigBriefInfo{
+					ID:        addr,
+					Signer:    signers,
+					Balance:   big.Min(big.Zero(), types.BigSub(act.Balance, locked)),
+					Threshold: threshold,
+				}
+				msigActorsInfo[addr] = info
+			}
+
+			if builtin.IsStorageMinerActor(act.Code) {
+				m, err := miner.Load(store, act)
+				if err != nil {
+					return xerrors.Errorf("fail to load miner actor: \n", err)
+				}
+
+				info, err := m.Info()
+				if err != nil {
+					return xerrors.Errorf("fail to get miner info: \n", err)
+				}
+
+				ownerId := lookupId(info.Owner)
+				if m, found := ownerMap[ownerId]; found { //add miner id to owner list
+					m = append(m, addr)
+					ownerMap[ownerId] = m
+				} else {
+					n := []address.Address{addr}
+					ownerMap[ownerId] = n
+				}
+
+				workerId := lookupId(info.Worker)
+				if m, found := workerMap[workerId]; found { //add miner id to worker list
+					m = append(m, addr)
+					workerMap[workerId] = m
+				} else {
+					n := []address.Address{addr}
+					workerMap[workerId] = n
+				}
+
+				bal, err := m.AvailableBalance(act.Balance)
```

Comment:
@arajasek @jennijuju I believe this is not enough. See the vested ( **not locked** ) here: https://github.com/filecoin-project/lotus/blob/v1.17.1/node/impl/full/state.go#L1310-L1320

Reply:
Thank youuuu!